galley/pkg/resource/synchronizer.go

```diff
@@ -0,0 +1,239 @@
+//  Copyright 2018 Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package resource
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"istio.io/istio/galley/pkg/change"
+	"istio.io/istio/pkg/log"
+)
+
+// Synchronizer is a Kubernetes controller that keeps a synchronized copy of an accessor set of custom
+// resources in destination APIGroup/Version(s). The mapping from sources to destinations are specified by the
+// respective schema.GroupVersion resources given during initialization.
+type Synchronizer struct {
+	// Lock for changing the running state of the accessor
+	stateLock sync.Mutex
+	// indicates whether the Synchronizer is running or not
+	running bool
+
+	// accessor for listening to, and operating on source and destination resources
+	source      *accessor
+	destination *accessor
+
+	// work queue for processing change events
+	queue workqueue.RateLimitingInterface
+
+	// wait until the background process task is completed
+	waitForProcess sync.WaitGroup
+
+	// event hook that gets called after every completion of event processing loop. Useful for testing
+	eventHook eventHookFn
+}
+
+type eventHookFn func(e interface{})
+
+// NewSynchronizer returns a new instance of a Synchronizer
+func NewSynchronizer(config *rest.Config, resyncPeriod time.Duration, name string, source schema.GroupVersion,
```

Comment:
This could be shorten a bit. Also the named return parameters aren't needed. e.g.

```golang
func NewSynchronizer(config *rest.Config, resyncPeriod time.Duration, name, kind, kindList string, src, dst schema.GroupVersion) (*Synchronizer,  error) {
	return newSynchronizer(config, resyncPeriod, name, source, destination, kind, listKind, nil)
```
}

Similar comment applies to line 68